faunadb-scala/src/test/scala/faunadb/CodecSpec.scala

```diff
@@ -0,0 +1,172 @@
+package faunadb
+
+import faunadb.values._
+import faunadb.values.time.HighPrecisionTime
+import org.joda.time.LocalDate
+import org.scalatest.{ FlatSpec, Matchers }
+
+class CodecSpec extends FlatSpec with Matchers {
+
+  // Encoding
+
+  it should "encode to primitive types" in {
+    LongV(10).to[Long].get shouldBe 10
+    LongV(10).to[Int].get shouldBe 10
+    LongV(10).to[Short].get shouldBe 10
+    LongV(10).to[Byte].get shouldBe 10
+    LongV(10).to[Char].get shouldBe 10
+
+    DoubleV(3.14).to[Double].get shouldBe 3.14
+    DoubleV(3.14).to[Float].get shouldBe 3.14f
+
+    BooleanV(true).to[Boolean].get shouldBe true
+
+    DateV("1970-01-03").to[LocalDate].get shouldBe LocalDate.parse("1970-01-03")
+
+    TimeV("1970-01-01T00:05:00.000000000Z").to[HighPrecisionTime].get shouldBe HighPrecisionTime.parse("1970-01-01T00:05:00.000000000Z")
+  }
+
+  it should "encode BytesV" in {
+    BytesV(1, 2, 3, 4).to[Array[Byte]].get shouldBe Array[Byte](1, 2, 3, 4)
+    BytesV(1, 2, 3, 4).to[String].get shouldBe "AQIDBA=="
+  }
+
+  case class Product(description: String, price: Double)
+  case class Order(customer: String, products: Seq[Product])
+
+  implicit val productCodec : Codec[Product] = Codec.custom[Product]
+  implicit val orderCodec : Codec[Order] = Codec.custom[Order]
+
+  it should "encode an object" in {
+    val product = ObjectV("description" -> StringV("laptop"), "price" -> DoubleV(999))
+
+    product.to[Product].get shouldBe Product("laptop", 999)
+  }
+
+  it should "encode nested objects" in  {
+    val product1 = ObjectV("description" -> StringV("laptop"), "price" -> DoubleV(999))
```

Comment:
Can you create one test where the order of keys doesn't match the order of the case class parameters?